### ISSUE TITLE
EnqueueRequestsFromMapFunc why add both new and old object into queue?

### ISSUE THREAD

**Issue by XBisATrouble** [kubernetes-sigs/controller-runtime#2276]
```go
// Update implements EventHandler
func (e *enqueueRequestsFromMapFunc) Update(evt event.UpdateEvent, q workqueue.RateLimitingInterface) {
	e.mapAndEnqueue(q, evt.ObjectOld)
	e.mapAndEnqueue(q, evt.ObjectNew)
}
``` 
why add both new and old object into queue?
Is it possible to just add new objects to queue?

**Comment by alvaroaleman** [comment]
Because we don't know which of the two matters. In 99% of cases, these two will be the same and we de-duplicate them.

If you need something else you will have to write a custom handler.

**Comment by seh** [comment]
> Is it possible to just add new objects to queue?

It's possible, but wrong. The mapping function could be sensitive to differences in the old and new views of object, causing it to project different reconciliation requests.

I have a controller that is sensitive to _Namespaces_ that match a particular label selector. I need to reconcile things both when I see a _Namespace_ whose labels _start_ matching that selector, and also when those labels _stop_ matching that selector. (The _Namespace_'s labels could change, and the selector can change over time.)

I rely on being able to see _Namespace_ objects both before and after such a change. If the _Namespace_ was matching the selector in its "old" view but no longer does in the "new" view, I don't have to worry about keeping track of which _Namespaces_ had matched the selector previously in order to notice that transition. Instead, my mapping function gets called on twice, and when it sees the old _Namespace_ matching the label selector, it emits some `reconcile.Request`s, and when it sees the new _Namespace_ no longer matching the label selector, it doesn't emit any `reconcile.Request`s. The result is that another controller winds up reconciling those `Request`s' target objects as intended.

**Comment by seh** [comment]
> Requiring the old state to correctly map is a problem, because there is no guarantee you will observe the state change.

Yes, I neglected to mention that possibility. I do have a fallback means of detecting that some objects related to a _Namespace_ that's no longer selected by the label selector are in need of reconciliation, but in most cases the normal event dispatching takes care of this detection.
